presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java

```diff
@@ -92,8 +92,10 @@ public PlanNode optimize(PlanNode plan, Session session, TypeProvider types, Pla
 
         Duration timeout = SystemSessionProperties.getOptimizerTimeout(session);
         Context context = new Context(memo, lookup, idAllocator, variableAllocator, System.nanoTime(), timeout.toMillis(), session, warningCollector);
-        exploreGroup(memo.getRootGroup(), context, matcher);
-
+        boolean planChanged = exploreGroup(memo.getRootGroup(), context, matcher);
```

Comment:
Why is this change needed? Why wasn't it needed before? If this is an independent optimization it feels like it is worth to extract it into a separate commit.

Reply:
I'm not sure if there are hidden reasons, but I suppose the reason that this is not needed previously is that the logicalPlanner does not care if a plan is actually rewritten or not by a specific rule, as long as the plan pass all the rules and finally become stable.

Now, for runtime optimization, we do care about whether the plan is actually rewritten or not. Because If not, we won't bother re-generating all the involved stageExecutionAndSchedulers, etc. Therefore, we make this change here so that if the plan is not actually rewritten, the original reference is returned, instead of a new identical plan. Doing so, we can easily tell (outside of iterativeOptimizer) if the returnedPlan is the unchanged by simply comparing the references.